File: csharp/GardenCatalogue-TestDesign-Kata/GardenCatalogue/GardenCatalogueManager.cs
namespace GardenCatalogue;

public class GardenCatalogueManager
{
    private readonly List<Plant> _plants = new();

    public GardenCatalogueManager()
    {
    }

    public GardenCatalogueManager(IEnumerable<Plant> plants)
    {
        _plants.AddRange(plants);
    }

    public int GetPlantCount() => _plants.Count;

    public Plant? GetPlantByName(string name) => _plants.FirstOrDefault(p => p.Name == name);

    public void AddPlant(Plant plant) => _plants.Add(plant);

    public IEnumerable<Plant> GetPlantsForCondition(SoilCondition soil, LightCondition light)
    {
        return _plants.Where(p => (p.Soil == soil || p.Soil == SoilCondition.Any) && p.Light == light);
    }

    public IEnumerable<Plant> PlanBed(Month month, double maxHeight)
    {
        return _plants.Where(p => p.Type == PlantType.Flower && 
                                 p.BloomPeriod.BloomsIn(month) && 
                                 p.MaxHeight <= maxHeight);
    }

    public IEnumerable<Plant> PlanHedge(double minHeight)
    {
        return _plants.Where(p => (p.Type == PlantType.Bush || p.Type == PlantType.Tree) && 
                                 p.MaxHeight >= minHeight);
    }
}

File: java/src/main/java/org/sammancoaching/GardenCatalogueManager.java
package org.sammancoaching;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class GardenCatalogueManager {
    private final List<Plant> plants;

    public GardenCatalogueManager() {
        this.plants = new ArrayList<>();
    }

    public GardenCatalogueManager(List<Plant> plants) {
        this.plants = new ArrayList<>(plants);
    }

    public void addPlant(Plant plant) {
        this.plants.add(plant);
    }

    public int getPlantCount() {
        return plants.size();
    }

    public List<Plant> getPlants() {
        return new ArrayList<>(plants);
    }

    public Plant getPlantByName(String name) {
        return plants.stream()
                .filter(p -> p.name().equals(name))
                .findFirst()
                .orElse(null);
    }

    public List<Plant> getPlantsForCondition(SoilCondition soil, LightCondition light) {
        return plants.stream()
                .filter(p -> (p.soil() == soil || p.soil() == SoilCondition.Any) && p.light() == light)
                .toList();
    }

    public List<Plant> planBed(Month month, double maxHeight) {
        return plants.stream()
                .filter(p -> p.type() == PlantType.Flower &&
                        p.bloomPeriod().bloomsIn(month) &&
                        p.maxHeight() <= maxHeight)
                .toList();
    }

    public List<Plant> planHedge(double minHeight) {
        return plants.stream()
                .filter(p -> (p.type() == PlantType.Bush || p.type() == PlantType.Tree) &&
                        p.maxHeight() >= minHeight)
                .toList();
    }
}
